README.md
# videojs-speak-descriptions-track

A Video.js 7 [middleware][middleware] that uses browser speech synthesis to speak descriptions contained in a description text track

If a text track with kind="descriptions" has mode="showing" in the player, then use speech synthesis (also known as "text to speech" or TTS) to announce each text cue when it is displayed. If it takes longer to announce the cue than its display time, then pause the tech until the speech synthesis finishes. Use a [middleware] layer to allow the tech to be paused without the player appearing to be paused, and to support pausing of the player either while both the tech and speech synthesis are playing, or while the tech is paused and the speech synthesis is still speaking.

Important Disclaimer: this plugin is not intended to be used to specifically address any accessibility requirements (e.g. WCAG 2.0 Success Criteria 1.2.3, 1.2.5, and 1.2.7). It is an experimental implementation of TTS for text video description. Users of this plugin should not assume that any level of accessibility support is provided by the features of this plugin.

## Table of Contents

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->


- [Installation](#installation)
- [Usage](#usage)
  - [`<script>` Tag](#script-tag)
  - [Browserify/CommonJS](#browserifycommonjs)
  - [RequireJS/AMD](#requirejsamd)
- [License](#license)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->
## Installation

```sh
npm install --save videojs-speak-descriptions-track
```

## Usage

To include videojs-speak-descriptions-track on your website or web application, use any of the following methods.
Since it's a middleware and attaches itself to Video.js automatically, it only needs to be included or required.

### `<script>` Tag

This is the simplest case. Get the script in whatever way you prefer and include the plugin _after_ you include [video.js][videojs], so that the `videojs` global is available.

```html
<script src="//path/to/video.min.js"></script>
<script src="//path/to/videojs-speak-descriptions-track.min.js"></script>
<script>
  var player = videojs('my-video');
</script>
```

### Browserify/CommonJS

When using with Browserify, install videojs-speak-descriptions-track via npm and `require` the plugin as you would any other module.

```js
var videojs = require('video.js');

// The actual middleware function is exported by this module, but it is also
// attached to the `Player.prototype`; so, there is no need to assign it
// to a variable.
require('videojs-speak-descriptions-track');

var player = videojs('my-video');
```

### RequireJS/AMD

When using with RequireJS (or another AMD library), get the script in whatever way you prefer and `require` the plugin as you normally would:

```js
require(['video.js', 'videojs-speak-descriptions-track'], function(videojs) {
  var player = videojs('my-video');
});
```

## License

MIT. Copyright (c) Owen Edwards &lt;owen.r.edwards@gmail.com&gt;

This video.js plugin is heavily based on https://github.com/videojs/videojs-playbackrate-adjuster, copyright (c) Gary Katsevman &lt;me@gkatsev.com&gt;

[videojs]: http://videojs.com/
[middleware]: https://docs.videojs.com/tutorial-middleware.html


CHANGELOG.md
<a name="1.7.0"></a>
# [1.7.0](https://github.com/OwenEdwards/videojs-speak-descriptions-track/compare/v1.6.0...v1.7.0) (2020-10-10)

<a name="1.6.0"></a>
# [1.6.0](https://github.com/OwenEdwards/videojs-speak-descriptions-track/compare/v1.5.0...v1.6.0) (2020-10-09)

<a name="1.5.0"></a>
# [1.5.0](https://github.com/OwenEdwards/videojs-speak-descriptions-track/compare/v1.4.0...v1.5.0) (2020-10-09)

<a name="1.4.0"></a>
# [1.4.0](https://github.com/OwenEdwards/videojs-speak-descriptions-track/compare/v1.3.1...v1.4.0) (2019-06-19)

<a name="1.3.1"></a>
## [1.3.1](https://github.com/OwenEdwards/videojs-speak-descriptions-track/compare/v1.3.0...v1.3.1) (2019-05-01)

### Chores

* run 'npm audit fix' to fix security vulnerabilities ([2caaece](https://github.com/OwenEdwards/videojs-speak-descriptions-track/commit/2caaece))

<a name="1.3.0"></a>
# [1.3.0](https://github.com/OwenEdwards/videojs-speak-descriptions-track/compare/v1.2.0...v1.3.0) (2019-05-01)

<a name="1.3.0"></a>
# [1.3.0](https://github.com/OwenEdwards/videojs-speak-descriptions-track/compare/v1.2.0...v1.3.0) (2019-05-01)

<a name="1.2.0"></a>
# [1.2.0](https://github.com/OwenEdwards/videojs-speak-descriptions-track/compare/v1.1.1...v1.2.0) (2019-04-02)
_(none)_

<a name="1.1.1"></a>
# [1.1.1](https://github.com/OwenEdwards/videojs-speak-descriptions-track/compare/v1.0.0...v1.1.1) (2019-03-15)
* Prevent text description track from being announced by a screen reader if it being announced by speech synthesis (4be9488)
* Fix poster URL in index.html to avoid browser warnings (747d795)
* Fix pause/play of the player during extended description (48fff6f)

<a name="1.0.0"></a>
# [1.1.1](https://github.com/OwenEdwards/videojs-speak-descriptions-track/compare/v1.0.0...v1.0.0) (2019-03-08)
* Initial Release


test/index.test.js
import {document} from 'global';

import QUnit from 'qunitjs';
import sinon from 'sinon';
import videojs from 'video.js';

import plugin from '../src/js/index.js';

const Player = videojs.getComponent('Player');

QUnit.module('sanity tests');

QUnit.test('the environment is sane', function(assert) {
  assert.strictEqual(typeof Array.isArray, 'function', 'es5 exists');
  assert.strictEqual(typeof sinon, 'object', 'sinon exists');
  assert.strictEqual(typeof videojs, 'function', 'videojs exists');
  assert.strictEqual(typeof plugin, 'function', 'plugin is a function');
});

QUnit.module('videojs-speak-descriptions-track', {

  beforeEach() {

    // Mock the environment's timers because certain things - particularly
    // player readiness - are asynchronous in video.js 5. This MUST come
    // before any player is created; otherwise, timers could get created
    // with the actual timer methods!
    this.clock = sinon.useFakeTimers();

    this.fixture = document.getElementById('qunit-fixture');
    this.video = document.createElement('video');
    this.fixture.appendChild(this.video);
    this.player = videojs(this.video);
  },

  afterEach() {
    this.player.dispose();
    this.clock.restore();
  }
});

QUnit.test('registers itself with video.js', function(assert) {
  assert.expect(2);

  assert.strictEqual(
    typeof Player.prototype.speakDescriptionsTrack,
    'function',
    'videojs-speak-descriptions-track plugin was registered'
  );

  this.player.speakDescriptionsTrack();

  // Tick the clock forward enough to trigger the player to be "ready".
  this.clock.tick(1);

  assert.ok(
    this.player.hasClass('vjs-playbackrate-adjuster'),
    'the plugin adds a class to the player'
  );
});


test/plugin.test.js
import document from 'global/document';

import QUnit from 'qunit';
import sinon from 'sinon';
import videojs from 'video.js';

import plugin from '../src/plugin';

// const Player = videojs.getComponent('Player');

QUnit.test('the environment is sane', function(assert) {
  assert.strictEqual(typeof Array.isArray, 'function', 'es5 exists');
  assert.strictEqual(typeof sinon, 'object', 'sinon exists');
  assert.strictEqual(typeof videojs, 'function', 'videojs exists');
  assert.strictEqual(typeof plugin, 'function', 'plugin is a function');
});

QUnit.module('videojs-speak-descriptions-track', {

  beforeEach() {

    // Mock the environment's timers because certain things - particularly
    // player readiness - are asynchronous in video.js 5. This MUST come
    // before any player is created; otherwise, timers could get created
    // with the actual timer methods!
    this.clock = sinon.useFakeTimers();

    this.fixture = document.getElementById('qunit-fixture');
    this.video = document.createElement('video');
    this.fixture.appendChild(this.video);
    this.player = videojs(this.video);
  },

  afterEach() {
    this.player.dispose();
    this.clock.restore();
  }
});

// QUnit.test('registers itself with video.js', function(assert) {
//   assert.expect(2);
//
//   assert.strictEqual(
//     typeof Player.prototype.speakDescriptionsTrack,
//     'function',
//     'videojs-speak-descriptions-track plugin was registered'
//   );
//
//   this.player.speakDescriptionsTrack();
//
//   // Tick the clock forward enough to trigger the player to be "ready".
//   this.clock.tick(1);
//
//   assert.ok(
//     this.player.hasClass('vjs-speak-descriptions-track'),
//     'the plugin adds a class to the player'
//   );
// });


src/js/index.js
import videojs from 'video.js';
import window from 'global/window';

/**
 * Player status for extended descriptions (playback of descriptions while pausing the tech)
 *
 * @typedef extendedPlayerState
 * @enum
 */
const extendedPlayerState = {
  unknown: 'unknown',
  initialized: 'initialized',
  playing: 'playing',
  paused: 'paused',
  playingExtended: 'playingExtended',
  pausedExtended: 'pausedExtended'
};

// TODO: user control over this attribute?
const audioDuckingFactor = 0.25;

/**
 * The SpeakDescriptionsTrackTTS component
 */
class SpeakDescriptionsTrackTTS {
  /**
   * Creates an instance of this class.
   *
   * @param {Player} player
   *        The `Player` that this class should be attached to.
   */
  constructor(player) {
    this.player_ = player;
    this.extendedPlayerState_ = extendedPlayerState.initialized;
    this.isDucked = false;

    if (window.speechSynthesis) {
      window.speechSynthesis.cancel();

      // Stop the textTrackDisplay component's element from having
      //  aria-live="assertive".
      let textTrackDisplay = player.getChild('textTrackDisplay');
      if (textTrackDisplay && textTrackDisplay.updateForTrack) {
        textTrackDisplay.originalUpdateForTrack = textTrackDisplay.updateForTrack;
        textTrackDisplay.updateForTrack = function(track) {
          if (this.getAttribute('aria-live') !== 'off') {
            this.setAttribute('aria-live', 'off');
          }
          this.originalUpdateForTrack(track);
        }.bind(textTrackDisplay);
      }
    }
  }

  /**
   * Dispose of the `SpeakDescriptionsTrackTTS`
   */
  dispose() {
  }

  play() {
    const speechSynthesis = window.speechSynthesis;

    if (speechSynthesis.paused) {
      speechSynthesis.resume();
    }
  }

  pause() {
    const speechSynthesis = window.speechSynthesis;

    if (speechSynthesis.speaking) {
      speechSynthesis.pause();
    }
  }

  paused() {
    return (
      this.extendedPlayerState_ === extendedPlayerState.paused ||
      this.extendedPlayerState_ === extendedPlayerState.pausedExtended
    );
  }

  textTrackChange() {
    const tracks = this.player_.textTracks();
    let descriptionsTrack = null;
    let i = tracks.length;

    while (i--) {
      const track = tracks[i];

      if (track.mode === 'showing') {
        if (track.kind === 'descriptions') {
          descriptionsTrack = track;
        }
      }
    }

    if (descriptionsTrack) {
      this.speakActiveCues(descriptionsTrack);
    }
  }

  /**
   * Use browser Speech Synthesis (aka TTS) to speak active cues, if supported
   *
   * @param {TextTrackObject} track Texttrack object to speak
   * @method speakActiveCues
   */
  speakActiveCues(track) {
    if (!window.SpeechSynthesisUtterance || !window.speechSynthesis) {
      return;
    }

    const speechSynthesis = window.speechSynthesis;

    let textToSpeak = [];
    let startTime = Infinity;
    let endTime = -Infinity;
    const ct = this.player_.currentTime();

    if (track.activeCues) {
      // TODO: Need to handle this logic better; it's possible that a new cue
      //       started while another is still active. We don't handle that correctly.
      for (let i = 0; i < track.activeCues.length; i++) {
        textToSpeak.push(track.activeCues[i].text);
        startTime = Math.min(track.activeCues[i].startTime, startTime);
        endTime = Math.max(track.activeCues[i].endTime, endTime);
      }
      // TODO: handle any HTML markup in the cues properly; for now,
      //       we just strip out HTML markup.
      textToSpeak = textToSpeak.join('\r\n').replace(/<(?:.|\n)*?>/gm, '');
    }

    if (textToSpeak) {
      if (speechSynthesis.speaking) {
        // TODO: Handle description cue collision
        videojs.log.warn(`Speech synthesis collision (${textToSpeak} - ${this.ssu.text}) : ${ct} : ${this.startTime} : ${this.endTime}`);

        speechSynthesis.cancel();

      } else if (speechSynthesis.paused) {
        // TODO: Handle if speech synthesis is paused here
        videojs.log.warn(`Speech synthesis collision (paused) (${textToSpeak} - ${this.ssu.text}) : ${ct} : ${this.startTime} : ${this.endTime}`);

        speechSynthesis.cancel();
        speechSynthesis.resume();
      }

      // Store info about the current cue for debugging and/or logging
      this.startTime = startTime;
      this.endTime = endTime;

      // TODO: Need to dispose of this ssu after it is finished?
      this.ssu = new window.SpeechSynthesisUtterance();

      this.ssu.text = textToSpeak;
      this.ssu.lang = this.increaseLanguageLocalization(track.language);

      // TODO: user control over these attributes
      this.ssu.rate = 1.1;
      this.ssu.pitch = 1.0;
      this.ssu.volume = 1.0;

      // TODO: This audio ducking needs to be made more robust
      this.ssu.onstart = function(e) {
        // Duck the player's audio
        if (!this.isDucked) {
          this.isDucked = true;
          this.player_.addClass('vjs-audio-ducked');
          this.player_.tech_.setVolume(this.player_.tech_.volume() * audioDuckingFactor);
        }
      }.bind(this);
      this.ssu.onend = function(e) {
        // Speech synthesis of a cue has ended

        const delta = (Date.now() - this.ssu.startDate) / 1000;

        videojs.log(`SpeakDescriptionsTrackTTS of cue: ${this.startTime} : ${this.endTime} : ${this.endTime - this.startTime} : ${delta} : ${(delta * 100.0 / (this.endTime - this.startTime)).toFixed(1)}%`);

        // Un-duck the player's audio
        if (this.isDucked) {
          this.isDucked = false;
          this.player_.removeClass('vjs-audio-ducked');
          this.player_.tech_.setVolume(this.player_.tech_.volume() / audioDuckingFactor);
        }

        if (this.extendedPlayerState_ === extendedPlayerState.playingExtended) {
          videojs.log('Un-pausing playback');
          this.extendedPlayerState_ = extendedPlayerState.playing;
          this.player_.tech_.play();
          this.descriptionExtended = false;
        }
      }.bind(this);
      this.ssu.onerror = function(e) {
        // An error occured during speech synthesis

        const delta = (Date.now() - this.ssu.startDate) / 1000;

        videojs.log.warn(`SSU error (${this.ssu.text})`);
        videojs.log.warn(`SpeakDescriptionsTrackTTS of cue: ${this.startTime} : ${this.endTime} : ${this.endTime - this.startTime} : ${delta} : ${(delta * 100.0 / (this.endTime - this.startTime)).toFixed(1)}%`);

        // Un-duck the player's audio
        if (this.isDucked) {
          this.isDucked = false;
          this.player_.removeClass('vjs-audio-ducked');
          this.player_.tech_.setVolume(this.player_.tech_.volume() / audioDuckingFactor);
        }

        if (this.extendedPlayerState_ === extendedPlayerState.playingExtended) {
          videojs.log('Un-pausing playback');
          this.extendedPlayerState_ = extendedPlayerState.playing;
          this.player_.tech_.play();
          this.descriptionExtended = false;
        }
      }.bind(this);

      // Start speaking the new textToSpeak

      this.ssu.startDate = Date.now();
      speechSynthesis.speak(this.ssu);

    } else {
      // No current textToSpeak, so a cue's display time has ended.

      if (speechSynthesis.speaking) {
        // Speech synthesis is still speaking - handle description cue overrun
        videojs.log('Pausing playback');

        this.extendedPlayerState_ = extendedPlayerState.playingExtended;
        this.descriptionExtended = true;
        this.player_.tech_.pause();

      } else if (speechSynthesis.paused) {
        // TODO: Handle if speech synthesis is paused here
        videojs.log.warn(`Speech synthesis overrun (paused) (${this.ssu.text}) : ${this.startTime} : ${this.endTime}`);

        speechSynthesis.cancel();
        speechSynthesis.resume();

   // } else if (this.ssu) {
     // videojs.log(`Speech had ended before end of cue (${this.ssu.text}) : ${this.startTime} : ${this.endTime} : ${ct}`);

      }

      return;
    }
  }

  /**
   * Try to improve the localization of the text track language, using
   *  the player's language setting and the browser's language setting.
   *  e.g. if lang='en' and language = 'en-US', use the more specific
   *  localization of language.
   *
   * @param {string} lang the lang attribute to try to improve
   * @return {string} the improved lang attribute
   * @method increaseLanguageLocalization
   */
  increaseLanguageLocalization(lang) {
    const playerLanguage = this.player_.language && this.player_.language();
    const navigatorLanguage = window.navigator && window.navigator.language;

    if (
      lang &&
      (typeof lang === 'string') &&
      (typeof playerLanguage === 'string') &&
      (playerLanguage.length > lang.length) &&
      (playerLanguage.toLowerCase().indexOf(lang.toLowerCase()) === 0)
    ) {

      lang = playerLanguage;
    }

    if (
      lang &&
      (typeof lang === 'string') &&
      (typeof navigatorLanguage === 'string') &&
      (navigatorLanguage.length > lang.length) &&
      (navigatorLanguage.toLowerCase().indexOf(lang.toLowerCase()) === 0)
    ) {

      lang = navigatorLanguage;
    }

    return lang;
  }
}

const speakDescriptionsTrack = function(player) {
  let tech;

  player.speakDescriptionsTTS = new SpeakDescriptionsTrackTTS(player);
  player.on('texttrackchange', player.speakDescriptionsTTS.textTrackChange.bind(player.speakDescriptionsTTS));
  player.on('dispose', player.speakDescriptionsTTS.dispose.bind(player.speakDescriptionsTTS));

  return {
    setSource(srcObj, next) {
      next(null, srcObj);
    },

    setTech(newTech) {
      tech = newTech;

      player.off(tech, 'pause', player.handleTechPause_);

      tech.on('pause', (event) => {
        if (player.speakDescriptionsTTS && player.speakDescriptionsTTS.extendedPlayerState_) {
          if (player.speakDescriptionsTTS.extendedPlayerState_ !== extendedPlayerState.playingExtended) {
            player.handleTechPause_();
          }
        }
      });
    },

    // TODO: Eventually we may modify the duration and/or current time to allow
    //       for the time that the video is paused for extended description.
    //       For now, we just treat it as though the video stalled while streaming.
    duration(dur) {
      return dur;
    },

    currentTime(ct) {
      return ct;
    },

    setCurrentTime(ct) {
      return ct;
    },

    volume(vol) {
      if (player.speakDescriptionsTTS && player.speakDescriptionsTTS.isDucked) {
        return vol / audioDuckingFactor;
      }

      return vol;
    },

    setVolume(vol) {
      if (player.speakDescriptionsTTS && player.speakDescriptionsTTS.isDucked) {
        return vol * audioDuckingFactor;
      }

      return vol;
    },

    paused() {
      if (player.speakDescriptionsTTS) {
        return player.speakDescriptionsTTS.paused();
      }
    },

    callPlay() {
      if (!player.speakDescriptionsTTS) {
        return;
      }

      if (!player.speakDescriptionsTTS.extendedPlayerState_) {
        player.speakDescriptionsTTS.extendedPlayerState_ = extendedPlayerState.unknown;
      }

      switch (player.speakDescriptionsTTS.extendedPlayerState_) {
      case extendedPlayerState.unknown:
      case extendedPlayerState.initialized:
      case extendedPlayerState.paused:
        player.speakDescriptionsTTS.extendedPlayerState_ = extendedPlayerState.playing;
        player.speakDescriptionsTTS.play();
        return;

      case extendedPlayerState.pausedExtended:
        player.speakDescriptionsTTS.extendedPlayerState_ = extendedPlayerState.playingExtended;
        player.speakDescriptionsTTS.play();
        player.handleTechPlay_();
        return videojs.middleware.TERMINATOR;
      }

      return;
    },

    callPause() {
      if (!player.speakDescriptionsTTS) {
        return;
      }

      if (!player.speakDescriptionsTTS.extendedPlayerState_) {
        player.speakDescriptionsTTS.extendedPlayerState_ = extendedPlayerState.unknown;
      }

      switch (player.speakDescriptionsTTS.extendedPlayerState_) {
      case extendedPlayerState.unknown:
      case extendedPlayerState.initialized:
      case extendedPlayerState.playing:
        player.speakDescriptionsTTS.extendedPlayerState_ = extendedPlayerState.paused;
        player.speakDescriptionsTTS.pause();
        return;

      case extendedPlayerState.playingExtended:
        player.speakDescriptionsTTS.extendedPlayerState_ = extendedPlayerState.pausedExtended;
        player.speakDescriptionsTTS.pause();
        player.handleTechPause_();
        return videojs.middleware.TERMINATOR;
      }

      return;
    }
  };
};

// Register the plugin with video.js.
videojs.use('*', speakDescriptionsTrack);

// Include the version number.
speakDescriptionsTrack.VERSION = '__VERSION__';

export default speakDescriptionsTrack;


src/plugin.js
import videojs from 'video.js';
import {version as VERSION} from '../package.json';
import window from 'global/window';

/**
 * Player status for extended descriptions (playback of descriptions while pausing the tech)
 *
 * @typedef extendedPlayerState
 * @enum
 */
const extendedPlayerState = {
  unknown: 'unknown',
  initialized: 'initialized',
  playing: 'playing',
  paused: 'paused',
  playingExtended: 'playingExtended',
  pausedExtended: 'pausedExtended'
};

// TODO: user control over this attribute?
const audioDuckingFactor = 0.25;

/**
 * The SpeakDescriptionsTrackTTS component
 */
class SpeakDescriptionsTrackTTS {
  /**
   * Creates an instance of this class.
   *
   * @param {Player} player
   *        The `Player` that this class should be attached to.
   */
  constructor(player) {
    this.player_ = player;
    this.extendedPlayerState_ = extendedPlayerState.initialized;
    this.isDucked = false;
    // TODO: user control over this setting
    this.originalSpeechRate = 1.1;
    this.speechRate = this.originalSpeechRate;

    if (window.speechSynthesis) {
      // workaround for chrome bug
      window.addEventListener('unload', () => {
        window.speechSynthesis.cancel();
        window.speechSynthesis.resume();
      });

      // Stop the textTrackDisplay component's element from having
      //  aria-live="assertive".
      const textTrackDisplay = player.getChild('textTrackDisplay');

      if (textTrackDisplay && textTrackDisplay.updateForTrack) {
        textTrackDisplay.originalUpdateForTrack = textTrackDisplay.updateForTrack;
        textTrackDisplay.updateForTrack = function(track) {
          if (this.getAttribute('aria-live') !== 'off') {
            this.setAttribute('aria-live', 'off');
          }
          this.originalUpdateForTrack(track);
        }.bind(textTrackDisplay);
      }
    }
  }

  voice(voice) {
    if (voice === undefined && this.voice_) {
      return this.voice_;
    } else if (Object.prototype.toString.call(voice) !== '[object SpeechSynthesisVoice]') {
      // reset to default voice;
      this.voice_ = null;

      const lang = this.ssu && this.ssu.lang || this.increaseLanguageLocalization(this.player_.language());

      return window.speechSynthesis.getVoices().filter(v => v.lang.startsWith(lang))[0];
    }

    this.voice_ = voice;
    return this.voice_;
  }

  /**
   * Dispose of the `SpeakDescriptionsTrackTTS`
   */
  dispose() {
  }

  play() {
    const speechSynthesis = window.speechSynthesis;

    // if (speechSynthesis.paused) {
    speechSynthesis.resume();
    // }
  }

  pause() {
    const speechSynthesis = window.speechSynthesis;

    if (speechSynthesis.speaking) {
      speechSynthesis.pause();
    }
  }

  paused() {
    return (
      this.extendedPlayerState_ === extendedPlayerState.paused ||
      this.extendedPlayerState_ === extendedPlayerState.pausedExtended
    );
  }

  textTrackChange() {
    const tracks = this.player_.textTracks();
    let descriptionsTrack = null;
    let i = tracks.length;

    while (i--) {
      const track = tracks[i];

      if (track.mode === 'showing') {
        if (track.kind === 'descriptions') {
          descriptionsTrack = track;
        }
      }
    }

    if (descriptionsTrack) {
      this.speakActiveCues(descriptionsTrack);
    }
  }

  /**
   * Use browser Speech Synthesis (aka TTS) to speak active cues, if supported
   *
   * @param {TextTrackObject} track Texttrack object to speak
   * @method speakActiveCues
   */
  speakActiveCues(track) {
    if (!window.SpeechSynthesisUtterance || !window.speechSynthesis) {
      return;
    }

    const speechSynthesis = window.speechSynthesis;

    let textToSpeak = [];
    let startTime = Infinity;
    let endTime = -Infinity;
    const ct = this.player_.currentTime();

    if (track.activeCues) {
      // TODO: Need to handle this logic better; it's possible that a new cue
      //       started while another is still active. We don't handle that correctly.
      for (let i = 0; i < track.activeCues.length; i++) {
        textToSpeak.push(track.activeCues[i].text);
        startTime = Math.min(track.activeCues[i].startTime, startTime);
        endTime = Math.max(track.activeCues[i].endTime, endTime);
      }
      // Replace newlines with spaces, since newlines are for visual layout,
      //  not to convey meaning.
      // TODO: handle valid HTML markup in the cues properly; for now,
      //       we just strip out HTML markup.
      textToSpeak = textToSpeak.join(' ').replace(/<(?:.|\n)*?>/gm, '').replace(/\r|\n/gm, ' ');
    }

    if (textToSpeak) {
      if (speechSynthesis.speaking) {
        // TODO: Handle description cue collision
        videojs.log.warn(`Speech synthesis collision (${textToSpeak} - ${this.ssu.text}) : ${ct} : ${this.startTime} : ${this.endTime}`);

        speechSynthesis.cancel();

      } else if (speechSynthesis.paused) {
        // TODO: Handle if speech synthesis is paused here
        videojs.log.warn(`Speech synthesis collision (paused) (${textToSpeak} - ${this.ssu.text}) : ${ct} : ${this.startTime} : ${this.endTime}`);

        speechSynthesis.cancel();
        speechSynthesis.resume();
      }

      // Store info about the current cue for debugging and/or logging
      this.startTime = startTime;
      this.endTime = endTime;

      // TODO: Need to dispose of this ssu after it is finished?
      this.ssu = new window.SpeechSynthesisUtterance();

      this.ssu.text = textToSpeak;
      this.ssu.lang = this.increaseLanguageLocalization(track.language);

      // get default voice for language or the user set voice
      this.ssu.voice = this.voice();

      // TODO: user control over these settings
      this.ssu.rate = this.player_.playbackRate() * this.speechRate;
      this.ssu.pitch = 1.0;
      this.ssu.volume = this.player_.volume();

      // TODO: This audio ducking needs to be made more robust
      this.ssu.onstart = this.duck.bind(this);
      this.ssu.onend = function(e) {
        // Speech synthesis of a cue has ended

        const delta = (Date.now() - this.ssu.startDate) / 1000;

        this.log({delta});

        // Adaptively change the speech rate to avoid repeated slight overruns
        const speechRatio = delta / (this.endTime - this.startTime);

        if (speechRatio > 1.0) {
          const newSpeechRate = this.speechRate * Math.sqrt(speechRatio);

          videojs.log(`Adjusting speech rate UP from ${this.speechRate} to ${newSpeechRate}`);
          this.speechRate = newSpeechRate;
        } else if ((speechRatio < 0.9) && (this.speechRate > this.originalSpeechRate)) {
          const newSpeechRate = (this.speechRate + this.originalSpeechRate) / 2.0;

          videojs.log(`Adjusting speech rate DOWN from ${this.speechRate} to ${newSpeechRate}`);
          this.speechRate = newSpeechRate;
        }

        this.utteranceFinished();
      }.bind(this);
      this.ssu.onerror = function(e) {
        // An error occured during speech synthesis

        const delta = (Date.now() - this.ssu.startDate) / 1000;

        videojs.log.warn(`SSU error (${this.ssu.text})`);
        this.log({delta, warn: true});

        this.utteranceFinished();
      }.bind(this);

      // Start speaking the new textToSpeak

      this.ssu.startDate = Date.now();
      speechSynthesis.speak(this.ssu);

    } else {
      // No current textToSpeak, so a cue's display time has ended.

      if (speechSynthesis.speaking) {
        // Speech synthesis is still speaking - handle description cue overrun
        videojs.log('Pausing playback');

        this.extendedPlayerState_ = extendedPlayerState.playingExtended;
        this.descriptionExtended = true;
        this.player_.tech_.pause();

      } else if (speechSynthesis.paused) {
        // TODO: Handle if speech synthesis is paused here
        videojs.log.warn(`Speech synthesis overrun (paused) (${this.ssu.text}) : ${this.startTime} : ${this.endTime}`);

        speechSynthesis.cancel();
        speechSynthesis.resume();

        // } else if (this.ssu) {
        // videojs.log(`Speech had ended before end of cue (${this.ssu.text}) : ${this.startTime} : ${this.endTime} : ${ct}`);

      }

      return;
    }
  }

  /**
   * Try to improve the localization of the text track language, using
   *  the player's language setting and the browser's language setting.
   *  e.g. if lang='en' and language = 'en-US', use the more specific
   *  localization of language.
   *
   * @param {string} lang the lang attribute to try to improve
   * @return {string} the improved lang attribute
   * @method increaseLanguageLocalization
   */
  increaseLanguageLocalization(lang) {
    const playerLanguage = this.player_.language && this.player_.language();
    const navigatorLanguage = window.navigator && window.navigator.language;

    if (
      lang &&
      (typeof lang === 'string') &&
      (typeof playerLanguage === 'string') &&
      (playerLanguage.length > lang.length) &&
      (playerLanguage.toLowerCase().indexOf(lang.toLowerCase()) === 0)
    ) {

      lang = playerLanguage;
    }

    if (
      lang &&
      (typeof lang === 'string') &&
      (typeof navigatorLanguage === 'string') &&
      (navigatorLanguage.length > lang.length) &&
      (navigatorLanguage.toLowerCase().indexOf(lang.toLowerCase()) === 0)
    ) {

      lang = navigatorLanguage;
    }

    return lang;
  }

  log({delta, warn = false}) {
    const log = warn ? videojs.log.warn : videojs.log;

    log(`SpeakDescriptionsTrackTTS of cue: ${this.startTime} : ${this.endTime} : ${this.endTime - this.startTime} : ${delta} : ${(delta * 100.0 / (this.endTime - this.startTime)).toFixed(1)}%`);
  }

  duck() {
    if (!this.isDucked) {
      this.isDucked = true;
      this.player_.addClass('vjs-audio-ducked');
      this.player_.tech_.setVolume(this.player_.tech_.volume() * audioDuckingFactor);
    }
  }

  unduck() {
    // Un-duck the player's audio
    if (this.isDucked) {
      this.isDucked = false;
      this.player_.removeClass('vjs-audio-ducked');
      this.player_.tech_.setVolume(this.player_.tech_.volume() / audioDuckingFactor);
    }
  }

  utteranceFinished() {
    this.unduck();

    if (this.extendedPlayerState_ === extendedPlayerState.playingExtended) {
      videojs.log('Un-pausing playback');
      this.extendedPlayerState_ = extendedPlayerState.playing;
      this.player_.tech_.play();
      this.descriptionExtended = false;
    }
  }
}

const speakDescriptionsTrack = function(player) {
  let tech;

  player.speakDescriptionsTTS = new SpeakDescriptionsTrackTTS(player);
  player.on('texttrackchange', player.speakDescriptionsTTS.textTrackChange.bind(player.speakDescriptionsTTS));
  player.on('dispose', player.speakDescriptionsTTS.dispose.bind(player.speakDescriptionsTTS));

  return {
    setSource(srcObj, next) {
      next(null, srcObj);
    },

    setTech(newTech) {
      tech = newTech;

      player.off(tech, 'pause', player.handleTechPause_);

      tech.on('pause', (event) => {
        if (player.speakDescriptionsTTS && player.speakDescriptionsTTS.extendedPlayerState_) {
          if (player.speakDescriptionsTTS.extendedPlayerState_ !== extendedPlayerState.playingExtended) {
            player.handleTechPause_();
          }
        }
      });
    },

    // TODO: Eventually we may modify the duration and/or current time to allow
    //       for the time that the video is paused for extended description.
    //       For now, we just treat it as though the video stalled while streaming.
    duration(dur) {
      return dur;
    },

    currentTime(ct) {
      return ct;
    },

    setCurrentTime(ct) {
      return ct;
    },

    volume(vol) {
      if (player.speakDescriptionsTTS && player.speakDescriptionsTTS.isDucked) {
        return vol / audioDuckingFactor;
      }

      return vol;
    },

    setVolume(vol) {
      if (player.speakDescriptionsTTS && player.speakDescriptionsTTS.isDucked) {
        return vol * audioDuckingFactor;
      }

      return vol;
    },

    paused() {
      if (player.speakDescriptionsTTS) {
        return player.speakDescriptionsTTS.paused();
      }
    },

    callPlay() {
      if (!player.speakDescriptionsTTS) {
        return;
      }

      if (!player.speakDescriptionsTTS.extendedPlayerState_) {
        player.speakDescriptionsTTS.extendedPlayerState_ = extendedPlayerState.unknown;
      }

      switch (player.speakDescriptionsTTS.extendedPlayerState_) {
      case extendedPlayerState.unknown:
      case extendedPlayerState.initialized:
      case extendedPlayerState.paused:
        player.speakDescriptionsTTS.extendedPlayerState_ = extendedPlayerState.playing;
        player.speakDescriptionsTTS.play();
        return;

      case extendedPlayerState.pausedExtended:
        player.speakDescriptionsTTS.extendedPlayerState_ = extendedPlayerState.playingExtended;
        player.speakDescriptionsTTS.play();
        player.handleTechPlay_();
        return videojs.middleware.TERMINATOR;
      }

      return;
    },

    callPause() {
      if (!player.speakDescriptionsTTS) {
        return;
      }

      if (!player.speakDescriptionsTTS.extendedPlayerState_) {
        player.speakDescriptionsTTS.extendedPlayerState_ = extendedPlayerState.unknown;
      }

      switch (player.speakDescriptionsTTS.extendedPlayerState_) {
      case extendedPlayerState.unknown:
      case extendedPlayerState.initialized:
      case extendedPlayerState.playing:
        player.speakDescriptionsTTS.extendedPlayerState_ = extendedPlayerState.paused;
        player.speakDescriptionsTTS.pause();
        return;

      case extendedPlayerState.playingExtended:
        player.speakDescriptionsTTS.extendedPlayerState_ = extendedPlayerState.pausedExtended;
        player.speakDescriptionsTTS.pause();
        player.handleTechPause_();
        return videojs.middleware.TERMINATOR;
      }

      return;
    }
  };
};

// Include the version number.
speakDescriptionsTrack.VERSION = VERSION;

// Register the plugin with video.js.
videojs.use('*', speakDescriptionsTrack);

export default speakDescriptionsTrack;
